File: 219. Contains Duplicate II/py/contains_nearby_duplicate.py
'''
LeetCode problem 219. Contains Duplicate II
https://leetcode.com/problems/contains-duplicate-ii/

Given an integer array `nums` and an integer `k`, return `true`
if there are two distinct indices `i` and `j` in the array
such that `nums[i] == nums[j]` and `abs(i - j) <= k`.
'''


def contains_nearby_duplicate(nums: list[int], k: int) -> bool:
    '''
    Runtime complexity: O(n)
    Space complexity: O(n)
    '''
    seen = {}
    for i, num in enumerate(nums):
        if num in seen and i - seen[num] <= k:
            return True
        seen[num] = i
    return False

File: 219. Contains Duplicate II/cs/ContainsNearbyDuplicate/ContainsNearbyDuplicate.cs
/*
LeetCode problem 219. Contains Duplicate II
https://leetcode.com/problems/contains-duplicate-ii/
*/

namespace LeetCode.ContainsNearbyDuplicate;

public class Solution
{
    public bool ContainsNearbyDuplicate(int[] nums, int k)
    {
        var seen = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            int num = nums[i];
            if (seen.TryGetValue(num, out int prevIndex) && i - prevIndex <= k)
                return true;
            seen[num] = i;
        }
        return false;
    }
}